### PR TITLE
Logins: Validate and fixup hostname and formSubmitURL.

### LINKED ISSUE #2157: Repair or omit invalid `formSubmitURL` and `hostname` values when querying the logins store
This is a spinoff from https://github.com/mozilla/application-services/issues/695.

For legacy reasons, various logins clients (including us!) have allowed invalid login data to be stored and synced. We need to get to a place where such invalid data is hidden from consuming applications, so that they don't get confused and make bad choices about what login data to autofill where.

This issue is to ensure that any invalid values already in the db or in sync, are not surfaced to the consuming application. When reading logins records from the db, we may need to:

* Validate that `hostname` is an origin.
* Validate that `formSubmitURL` is an origin, if present.
* Probably some other stuff here, depending on the output of https://github.com/mozilla/application-services/issues/2155

Note that this work may be redundant in the face of https://github.com/mozilla/application-services/issues/695, which is about *fixing up* invalid records in the db. But it seems like it could be independently useful until we're sure that's working correctly in all cases.

┆Issue is synchronized with this [Jira Story](https://jira.mozilla.com/browse/SYNC-355)
┆Epic: important not urgent

### PR DESCRIPTION
**Pull request by mhammond**
Also tackles most TODO items for these fields, plus some other tests for good measure

However, the tests fail due to an issue with expecting `file:///` to be a valid hostname - we always remove the trailing "/" (ie, we insist there's no path component) so according to that rule `file://` is what should be considered valid. I probably need to chat with @mnoorenberghe about this.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `cargo test --all` produces no test failures
  - `cargo clippy --all --all-targets --all-features` runs without emitting any warnings
  - `cargo fmt` does not produce any changes to the code
  - `./gradlew ktlint detekt` runs without emitting any warnings
  - `swiftformat --swiftversion 4 megazords components/*/ios && swiftlint` runs without emitting any warnings or producing changes
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/logins/src/login.rs [hunk 1]
```diff
@@ -112,15 +96,8 @@
 //!   - replacing invalid values with null if a valid 'httpRealm' field is present
 //!
 //!   **XXX TODO**:
-//!   - test that we validate as an origin.
-//!   - test that we fixup full URLs by turning them into origins?
-//!   - test that we fixup "." to the empty string
```
Comment (lougeniaC64) [inline components/logins/src/login.rs]: I don't think we fixup for this case (yet?).

**Comment by mhammond** [comment]
Actually, this doesn't fix #2157, as that calls for fixing up at *query* time - but the changes here fixup at insert time. The stuff introduced here would be suitable for fixing up the DB as part of a "migration" as suggested in #2153

**Comment by rfk** [comment]
I believe this fixup logic does currently get called at query time, by virtue of `Login::from_row`.

**Comment by mhammond** [comment]
> I believe this fixup logic does currently get called at query time, by virtue of `Login::from_row`.

Right - so I guess this means that so long as we don't locate stuff via sql, we should be good. However, our dupe logic does exactly that (via `dupe_exists()`) so it's not clear to me things will actually work as expected for non-normalized URLs already in the DB. I'll try and knock up a test case to (dis-)prove this. In the meantime I'm merging this, and we can close #2157 manually if we decide that makes sense.

**Pull request merged**